### PR TITLE
feat: impl send_transaction

### LINKED ISSUE #1768: Impl `send_transaction`
### Describe the feature

Now that we have signing #1665 

we can impl `send_transaction`

# TODO

impl:

https://github.com/paradigmxyz/reth/blob/bba61c0b617e4d2f9569b1524c583431792949ad/crates/rpc/rpc/src/eth/api/transactions.rs#L144-L144

we need 
1. Conversion from `TransactionRequest -> TypedTransactionRequest`, see also for reference: https://github.com/foundry-rs/foundry/blob/bb62e3c8834625882ffc114ac14c78d1db213a0f/anvil/core/src/eth/transaction/mod.rs#L84-L84
2. Convert `TypedTransactionRequest` to `primitive::Transaction` https://github.com/paradigmxyz/reth/blob/bba61c0b617e4d2f9569b1524c583431792949ad/crates/primitives/src/transaction/mod.rs#L170-L170 (both steps could be 1 step and bypass the `TypedTransactionRequest` intermediate step)

3. impl https://github.com/paradigmxyz/reth/blob/bba61c0b617e4d2f9569b1524c583431792949ad/crates/rpc/rpc/src/eth/signer.rs#L73-L79

4. submit to pool

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by chirag-bgh**
Closes #1768

### CODE REVIEW

**Comment by chirag-bgh** [comment]
@mattsse To make sure if I'm proceeding the right way. 

For the rlp encoding of tx needed in `sign_transaction`, we need to define these functions in `reth-rlp`  right? https://github.com/gakonst/ethers-rs/blob/304fe0988d23302694a63da47ce683af19305357/ethers-core/src/types/transaction/request.rs#L162-L175

https://github.com/gakonst/ethers-rs/blob/304fe0988d23302694a63da47ce683af19305357/ethers-core/src/types/transaction/mod.rs#L20

**Comment by mattsse** [comment]
ptal at the referenced anvil code.

for reth this means, 

1. we need to go from `TypedTransactionRequest` to

https://github.com/paradigmxyz/reth/blob/2a457c70238256b1fcf00ab1683500281576468c/crates/primitives/src/transaction/mod.rs#L170-L170

filling any missing fields

2. get the Signature by signing the transaction's signature hash (see Signer trait `sign_message`)
3. Wrap into `TransactionSigned` https://github.com/paradigmxyz/reth/blob/2a457c70238256b1fcf00ab1683500281576468c/crates/primitives/src/transaction/mod.rs#L655-L655
4. submit into pool

File: crates/rpc/rpc-types/src/eth/transaction/typed.rs [hunk 1]
```diff
@@ -112,3 +112,12 @@ impl Decodable for TransactionKind {
         }
     }
 }
+
+impl From<TransactionKind> for reth_primitives::TransactionKind {
```
Comment (gakonst) [inline crates/rpc/rpc-types/src/eth/transaction/typed.rs:116]: @mattsse I assume this type won't be necessary in the future and we'll be using it from Revm? here we implemented it again just for the RLP methods?

File: crates/rpc/rpc/src/eth/api/transactions.rs [hunk 2]
```diff
@@ -152,8 +154,43 @@ where
...
+    /// This will sign the transaction and submit it to the pool
     pub(crate) async fn send_transaction(&self, _request: TransactionRequest) -> EthResult<H256> {
-        unimplemented!()
+        let from = _request.from.unwrap_or_default();
+        let _transaction = match _request.into_typed_request() {
+            Some(tx) => tx,
+            None => return Err(EthApiError::Unsupported("invalid transaction type")),
+        };
+
+        let signed_tx = self.sign_request(&from, _transaction)?;
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/api/transactions.rs:167]:
```suggestion
        let transaction = match request.into_typed_request() {
            Some(tx) => tx,
            None => return Err(EthApiError::Unsupported("invalid transaction type")),
        };

        let signed_tx = self.sign_request(&from, transaction)?;
```

File: crates/rpc/rpc/src/eth/api/transactions.rs [hunk 3]
```diff
@@ -152,8 +154,43 @@ where
     Client: BlockProvider + StateProviderFactory + EvmEnvProvider + 'static,
     Network: 'static,
 {
+    /// Send a transaction to the pool
+    ///
+    /// This will sign the transaction and submit it to the pool
     pub(crate) async fn send_transaction(&self, _request: TransactionRequest) -> EthResult<H256> {
-        unimplemented!()
+        let from = _request.from.unwrap_or_default();
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/api/transactions.rs:161]: Is this right? I believe Geth requires that `from` is specified?

File: crates/rpc/rpc/src/eth/api/transactions.rs [hunk 4]
```diff
@@ -152,8 +154,43 @@ where
...
+        Ok(hash)
+    }
+
+    pub(crate) fn sign_request(
+        &self,
+        from: &Address,
+        request: TypedTransactionRequest,
+    ) -> EthResult<TransactionSigned> {
+        for signer in self.inner.signers.iter() {
+            if signer.accounts().contains(from) {
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/api/transactions.rs:186]: nit https://github.com/paradigmxyz/reth/blob/c110ede505e3f490863f0225c2909eb78030904a/crates/rpc/rpc/src/eth/signer.rs#L22

File: crates/rpc/rpc/src/eth/signer.rs [hunk 5]
```diff
@@ -75,7 +79,46 @@ impl EthSigner for DevSigner {
         _request: TypedTransactionRequest,
         _address: &Address,
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/signer.rs:80]:
```suggestion
        request: TypedTransactionRequest,
        address: &Address,
```

File: crates/rpc/rpc/src/eth/signer.rs [hunk 6]
```diff
@@ -75,7 +79,46 @@ impl EthSigner for DevSigner {
...
     ) -> Result<TransactionSigned> {
-        todo!()
+        // convert to primitive transaction
+        let mut _transaction = match _request {
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/signer.rs:83]: Do we not have a utility for converting across types?

**Review by gakonst — APPROVED** [review]
fine w me - pls use dont use unerscored variable names for things that aren't left temporarily unused. let's also dedup the tx conversion type

File: crates/rpc/rpc/src/eth/signer.rs (hunk 6 above)
Comment (gakonst) [reply crates/rpc/rpc/src/eth/signer.rs]: please add one!

Comment (chirag-bgh) [reply crates/rpc/rpc/src/eth/signer.rs]: done!

**Review by mattsse — APPROVED** [review]
tysm lgtm,

there are a few things we need to do, some of it are currently blocked by missing gas_price function for example.

will open another issue for this.

**Pull request merged**